ReSwift/CoreTypes/Subscription.swift

```diff
@@ -8,159 +8,87 @@
+        _initialState = nil
 
-    init() {}
+        return initialState
+    }
 
-    /// Initializes a subscription with a sink closure. The closure provides a way to send
-    /// new values over this subscription.
-    private init(sink: @escaping (@escaping (State?, State) -> Void) -> Void) {
-        // Provide the caller with a closure that will forward all values
-        // to observers of this subscription.
-        sink { old, new in
-            self.newValues(oldState: old, newState: new)
+    public func select<SelectedState>(_ selector: @escaping (State) -> SelectedState) -> Subscription<SelectedState> {
+        let subscription = Subscription<SelectedState>(
+            initialState: selector(self.initialState()),
+            automaticallySkipsEquatable: automaticallySkipsEquatable
+        )
+        notify = { oldState, newState in
+            let newSelected = selector(newState)
+            let oldSelected = selector(oldState)
+            subscription.notify(oldSelected, newSelected)
         }
+        return subscription
     }
 
-    /// Sends new values over this subscription. Observers will be notified of these new values.
-    func newValues(oldState: State?, newState: State) {
-        self.observer?(oldState, newState)
+    public func skip(when skip: @escaping (_ oldState: State, _ newState: State) -> Bool) -> Subscription<State> {
+        let subscription = Subscription<State>(
+            initialState: self.initialState(),
+            automaticallySkipsEquatable: false // Since we're declaring a skip manually, we can avoid double-skips
+        )
+        notify = { oldState, newState in
+            if !skip(oldState, newState) {
+                subscription.notify(oldState, newState)
+            }
+        }
+        return subscription
     }
 
-    /// A caller can observe new values of this subscription through the provided closure.
-    /// - Note: subscriptions only support a single observer.
-    fileprivate func observe(observer: @escaping (State?, State) -> Void) {
-        self.observer = observer
+    public func only(when only: @escaping (_ oldState: State, _ newState: State) -> Bool) -> Subscription<State> {
+        let subscription = Subscription<State>(
+            initialState: self.initialState(),
+            automaticallySkipsEquatable: false // Since we're declaring a skip manually, we can avoid double-skips
+        )
+        notify = { oldState, newState in
+            if only(oldState, newState) {
+                subscription.notify(oldState, newState)
+            }
+        }
+        return subscription
     }
-}
 
-extension Subscription where State: Equatable {
-    public func skipRepeats() -> Subscription<State>{
-        return self.skipRepeats(==)
+    public func subscribe<S: StoreSubscriber>(_ subscriber: S) where S.StoreSubscriberStateType == State {
+        assert(self.subscriber == nil, "Subscriptions do not support multiple subscribers")
```

Comment:
See comment above; not allowing multiple calls to subscribe() is unintuitive and makes the function a partially applied function, if I remember the term correctly, through the (debug build-only) assertion. I think this should be a precondition that affects client code, because if not, the behavior is supposed to be undefined.
 but will not show any sign of a problem.

That being said, in accordance to the stuff above, I think it should be subscribe(store:) instead, with a PendingSubscription from store.subscription() being sugar to hold on to a Store and Subscription instance, and PendingSubscription.subscribe() simply passing the store to the subscription.